fix(mechanics): Store Outfit attributes as integers to avoid precision errors when selling large numbers of outfits

Amazinite opened the pull request:
**Bug fix**

This PR addresses the bug described in issue #3639.
Fixes #3639.

## Acknowledgement

- [x] I acknowledge that I have read and understand the [Contributing](https://github.com/endless-sky/endless-sky/blob/master/docs/CONTRIBUTING.md) article.

## Summary

This PR changes the Dictionary class to be a template class, allowing control over what the Dictionary stores. The Outfit class now contains two dictionaries, a double dictionary and an int64_t dictionary. When loading outfits, determining if you can buy/sell an outfit, and when adding/removing outfits to/from a ship, the int64_t dictionary is used, ensuring that there are no problems caused by floating point precision errors when adding and removing outfits. For all other purposes, the double dictionary is used to maintain the same functionality for classes that use Outfit.

As a result of this, attributes can now only have values as small as 0.0001.

## Testing Done

Bought 300,001 grenades. Reloaded the game, then tried to sell all of them. Without this PR, I get stuck with 1 grenade I can't sell unless I reload the save. With it, I can sell every grenade without issue.

## Performance Impact

This on its own actually reduces performance. #12027 addresses this performance hit.

Amazinite committed:
fix(mechanics): Store Outfit attributes as integers to avoid precision errors when selling large numbers of outfits

Amazinite committed:
Feedback from Tibetiroka + fix weapon ports

Amazinite committed:
Please Microsoft

Amazinite committed:
Remove std prefixes

Amazinite commented on source/Outfit.cpp:
```diff
@@ -592,7 +592,7 @@ void Outfit::AddLicenses(const Outfit &other)
 void Outfit::Set(const char *attribute, double value)
 {
-	attributes[attribute] = value;
+	rawAttributes[attribute] = value * ATTRIBUTE_PRECISION;
```
```suggestion
	rawAttributes[attribute] = value * ATTRIBUTE_PRECISION;
	cachedAttributes[attribute] = value;
```

Amazinite committed:
Update both Dictionaries when calling Outfit::Set

petervdmeer commented on source/Outfit.h:
```diff
@@ -140,7 +142,8 @@ class Outfit {
 
-	Dictionary attributes;
+	Dictionary<int64_t> rawAttributes;
+	Dictionary<double> cachedAttributes;
```
It makes a lot of sense to me to put cachedAttributes on ships realtime in-flight, because we don't want to have the performance overhead of converting int64_t to double when the player is in-flight.
I see less of a need to have this cache directly on outfits. Are there other situations than ships realtime in-flight where it is desired that outfits also have this cache? (If not, then I would expect outfits to only contain the int64_t version of the data.)

Amazinite replied:
While we're here, I wonder if it's time for me to finally resurrect #6804. That PR was basically an attribute cache. It was previously done to avoid string comparisons during the attribute lookups, but now it also dodges a division, so it's probably even more beneficial now than it was before.

petervdmeer commented on source/Outfit.cpp:
```diff
@@ -30,7 +30,9 @@ this program. If not, see <https://www.gnu.org/licenses/>.
-	const double EPS = 0.0000000001;
+	// Attributes are stored as integers but used as doubles. This is the factor by which to convert
+	// attribute values from doubles to integers and back.
+	constexpr int ATTRIBUTE_PRECISION = 10000;
```
```suggestion
	// Attributes are stored as integers but used as doubles. This is the factor by which to convert
	// attribute values from doubles to integers and back. Using 0xffff (65535), because powers of
	// two are faster to multiply and divide than powers of ten.
	constexpr int ATTRIBUTE_PRECISION = 0xffff;
```

Amazinite commented on source/Dictionary.h:
```diff
@@ -25,19 +25,21 @@ this program. If not, see <https://www.gnu.org/licenses/>.
 // compared to an STL map. That makes it suitable for ship attributes, which are
 // changed much less frequently than they are queried.
-class Dictionary : private std::vector<std::pair<const char *, double>> {
+template<class Type>
```
The three lines at the bottom of the CPP files are the explicit instantiations.

Amazinite committed:
Don't cache the double values for attributes within Outfit

Performance improvements will come later

Amazinite committed:
Replace the Outfit::AttributeNames function with iterator functions over the attribute names and values

Amazinite commented on source/Outfit.h:
```diff
@@ -71,9 +97,14 @@ class Outfit {
 	const Sprite *Thumbnail() const;
 
+	// Return true if this Outfit's attributes Dictionary is empty. Does not determine
+	// whether this Outfit contains any sprites, effects, sounds.
```
```suggestion
	// whether this Outfit contains any sprites, effects, sounds, or a weapon.
```

Amazinite committed:
English

Amazinite committed:
Merge remote-tracking branch 'upstream/master' into outfit-attribute-precision

Amazinite committed:
Fix merge

Amazinite committed:
Merge remote-tracking branch 'upstream/master' into outfit-attribute-precision

Amazinite committed:
Merge remote-tracking branch 'upstream/master' into outfit-attribute-precision

Amazinite committed:
Use precise values when uninstalling linked outfits + guard against pushing an outfit into the negative

Amazinite committed:
Fix uninstall counting

Amazinite committed:
Housekeeping

Use precise values in a few locations where we only care about whether the value is positive, negative, or 0, since it skips a division.
